Change default level of INVALID_HTML_TAGS to warning and stabilize it

GuillaumeGomez opened the pull request:
Fixes of #67799.

cc @Nemo157 
r? @notriddle

GuillaumeGomez committed:
Stabilize rustdoc CHECK_INVALID_HTML_TAGS check

GuillaumeGomez committed:
Change default lint level of INVALID_HTML_TAGS to warning

Nemo157 commented on library/alloc/src/sync.rs:
```diff
@@ -69,42 +69,42 @@ macro_rules! acquire {
-/// allocated in the heap. Invoking [`clone`][clone] on `Arc` produces
-/// a new `Arc` instance, which points to the same allocation on the heap as the
-/// source `Arc`, while increasing a reference count. When the last `Arc`
+/// The type [`Arc`]`<T>` provides shared ownership of a value of type `T`,
```
This will currently render quite badly: https://github.com/rust-lang/rust/issues/62867

GuillaumeGomez replied:
Ah right. We can mitigate this problem by adding this CSS:

```css
.docblock code + a > code {
    padding: 0;
}
```

But it doesn't fix the right padding on the link. Still better than nothing though I guess.

GuillaumeGomez replied:
The link would be on the whole `Arc<T>`. I replicated what was there originally. It's done like that in a lot of places.

Nemo157 commented on library/alloc/src/vec/mod.rs:
```diff
@@ -296,7 +296,7 @@ mod spec_extend;
 /// types inside a `Vec`, it will not allocate space for them. *Note that in this case
 /// the `Vec` might not report a [`capacity`] of 0*. `Vec` will allocate if and only
-/// if <code>[mem::size_of::\<T>]\() * [capacity]\() > 0</code>. In general, `Vec`'s allocation
+/// if [`mem::size_of`]`::<T>() * `[`capacity`]`() > 0`. In general, `Vec`'s allocation
```
This seems to be a false-positive from the lint. If I copy the original code into a local project and build the docs there is no `<T>` tag in the output.

GuillaumeGomez replied:
Considering there were hundreds of links, I mostly looked for `<code>` tags and replaced. So a few of them just got an update because of that.

Nemo157 replied:
I mean I also tested the lint locally and it warned on this:

```
warning: unclosed HTML tag `T`
 --> src/lib.rs:4:30
  |
4 | /// if <code>[mem::size_of::\<T>]\() * [capacity]\() > 0</code>.
  |                              ^^^
```

GuillaumeGomez replied:
Oh indeed. Interesting. However, this is apparently the expected behaviour since we even enforce it through a test: https://github.com/rust-lang/rust/blob/master/src/test/rustdoc-ui/intra-doc/malformed-generics.rs

GuillaumeGomez replied:
Hopefully I will be able to remove this whole commit.

Nemo157 commented:
Most of the doc changes seem to basically be a direct revert of https://github.com/rust-lang/rust/pull/88343

GuillaumeGomez commented:
> Most of the doc changes seem to basically be a direct revert of #88343

Unfortunately yes. It should have been fixed in the doc generation, not like this.

GuillaumeGomez commented:
I'll rebase on top of https://github.com/rust-lang/rust/pull/101731 once merged then integrate the stabilization part of the PR from [here](https://github.com/rust-lang/rust/compare/master...GuillaumeGomez:rust:stabilize-CHECK_INVALID_HTML_TAGS?expand=1) then start the FCP.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3038698518/jobs/4892734936) [(plain)](https://github.com/rust-lang/rust/commit/79bbd1c5d992e076f219f43f8fb70737dde34113/checks/8309848840/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Rustbook (x86_64-unknown-linux-gnu) - embedded-book
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.16s
core/option/index.html:42: broken intra-doc link - [<code>Box&lt;T&gt;</code>]
std/option/index.html:42: broken intra-doc link - [<code>Box&lt;T&gt;</code>]
number of HTML files scanned: 32982
number of HTML redirects found: 10114
number of links checked: 2474532
number of links ignored due to external: 118069
```

</details>

GuillaumeGomez committed:
Update rustdoc tests

GuillaumeGomez commented:
## Stabilization Report

### Implementation History

- Initial implementation (https://github.com/rust-lang/rust/pull/77119).
- Expanded to HTML comments (https://github.com/rust-lang/rust/pull/77753).
- Added suggestions to turn the unclosed HTML tag into a codeblock (https://github.com/rust-lang/rust/pull/93569).
- Fixed the duplicated warnings with the intra-doc link features (https://github.com/rust-lang/rust/pull/93605).
- Set the lint level to warn and stabilize it (https://github.com/rust-lang/rust/pull/101720)

### Summary

This lint warns about unclosed HTML tags and comments. See https://doc.rust-lang.org/nightly/rustdoc/lints.html#invalid_html_tags for more information.

### Experience Report

Before this lint was adding, an unclosed HTML tag could lead to broken generated documentation appearance without providing a way for users to know what went wrong.

GuillaumeGomez commented:
Then let's start the FCP process to hear everyone's opinion about this. I think it should be at warn level by default before the stabilization because after would finally be the same problem as the new warning would appear from nowhere for users.

Nominating for discussion. Also kicking off an FCP: shall we make it a warning before or after stabilization?

@rfcbot merge

CraftSpider commented:
Changing to warn before stabilization seems fine to me, as stable doesn't disallow typing `#[allow(rustdoc::invalid_html_tags)]`. People getting the warning on nightly-only briefly, then get it on both shortly after, seems reasonable. Either way, I'm for stabilizing the lint.

camelid commented on src/test/rustdoc-ui/intra-doc/malformed-generics.stderr:
```diff
@@ -23,80 +23,136 @@ LL | //! [Vec<Box<T>]
+LL | //! [Vec<Box<T>]
+   |             ^^^
+   |
+   = note: `#[warn(rustdoc::invalid_html_tags)]` on by default
```
@rfcbot concern intra-doc link suggestion

Hmm, this should be suggesting wrapping the contents of the link with ` `` `. I've seen people writing links like `[Vec<T>]` before, so we should definitely be giving them a solution to the lint (that's also the standard to do things).

camelid commented:
@rfcbot concern intra-doc link suggestion

See https://github.com/rust-lang/rust/pull/101720#discussion_r971258140.

GuillaumeGomez commented:
@camelid Did [this comment](https://github.com/rust-lang/rust/pull/101720#discussion_r971276777) answer your concern?

camelid commented:
Yeah, looks good. I hadn't noticed the mismatched brackets.

camelid commented:
@rfcbot resolve intra-doc link suggestion

bors commented:
:pushpin: Commit f92d2944bd0c56a740a89a206ece5c85bcc215b4 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle f92d2944bd0c56a740a89a206ece5c85bcc215b4 -->
<!-- homu: {"type":"Approved","sha":"f92d2944bd0c56a740a89a206ece5c85bcc215b4","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- failed in a rollup 
https://github.com/rust-lang-ci/rust/actions/runs/3210249920/jobs/5247560165#step:25:32344
https://github.com/rust-lang/rust/pull/102802

GuillaumeGomez committed:
Fix doc lint error

GuillaumeGomez commented:
At least it proves that the lint is working as expected. :rofl: I pushed a fix for it.

GuillaumeGomez commented:
@bors r=notriddle

bors commented:
:pushpin: Commit 4ee7db23b6b8fa9f8a66b4f378387715c95a5b86 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 4ee7db23b6b8fa9f8a66b4f378387715c95a5b86 -->
<!-- homu: {"type":"Approved","sha":"4ee7db23b6b8fa9f8a66b4f378387715c95a5b86","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez committed:
Fix compiler docs

GuillaumeGomez commented:
I wouldn't say bitrot, more like a lot of docs were added to the compiler (which is great!). The lint allows us to see them. :smile:

bors commented:
:pushpin: Commit a01eb1cdf9e6e125d9ff876fe37fceab62dece51 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle a01eb1cdf9e6e125d9ff876fe37fceab62dece51 -->
<!-- homu: {"type":"Approved","sha":"a01eb1cdf9e6e125d9ff876fe37fceab62dece51","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit a01eb1cdf9e6e125d9ff876fe37fceab62dece51 with merge 24100e3af2c9f14a5f26536601f00b08ebf0c147...
<!-- homu: {"type":"BuildStarted","head_sha":"a01eb1cdf9e6e125d9ff876fe37fceab62dece51","merge_sha":"24100e3af2c9f14a5f26536601f00b08ebf0c147"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3220619107/jobs/5268022673)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3220619107/jobs/5268022673","builder_name":"checks-actions"} -->

GuillaumeGomez committed:
Fix unclosed HTML tag in rustfmt doc

bors commented:
:pushpin: Commit 14de94aec59bcd535f690e255a950ca76fcf2d5a has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 14de94aec59bcd535f690e255a950ca76fcf2d5a -->
<!-- homu: {"type":"Approved","sha":"14de94aec59bcd535f690e255a950ca76fcf2d5a","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3220619107/jobs/5267552161) [(plain)](https://github.com/rust-lang-ci/rust/commit/24100e3af2c9f14a5f26536601f00b08ebf0c147/checks/8804849434/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc-workspace-hack v1.0.0 (/checkout/src/tools/rustc-workspace-hack)
[RUSTC-TIMING] rustc_workspace_hack test:false 0.034
[RUSTC-TIMING] cargo_metadata test:false 1.534
 Documenting rustfmt-nightly v1.5.1 (/checkout/src/tools/rustfmt)
error: unclosed HTML tag `type`
  |
  |
7 |     /// pipe-separated list of variants; for other types it returns "<type>".
  |
  |
  = note: `-D rustdoc::invalid-html-tags` implied by `-D warnings`
error: could not document `rustfmt-nightly`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name rustfmt_nightly src/tools/rustfmt/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/doc --cfg 'feature="cargo-fmt"' --cfg 'feature="default"' --cfg 'feature="rustfmt-format-diff"' -Zunstable-options --check-cfg 'values(feature, "cargo-fmt", "default", "generic-simd", "rustfmt-format-diff")' --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=0952539fdab0a963 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps --extern annotate_snippets=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libannotate_snippets-3d01d70feecb4c16.rmeta --extern anyhow=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libanyhow-d2aa2f66548eaa6c.rmeta --extern bytecount=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libbytecount-08afd061e326b889.rmeta --extern cargo_metadata=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libcargo_metadata-afa6427d7f2b495c.rmeta --extern clap=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclap-a920ce44725b3495.rmeta --extern derive_new=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libderive_new-52a645b5b90627c6.so --extern diff=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libdiff-ee868c24f9c843f8.rmeta --extern dirs=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libdirs-5fe7fdae19ee57b4.rmeta --extern env_logger=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libenv_logger-2d5cf7581501a598.rmeta --extern getopts=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libgetopts-3aacf0710af4af4e.rmeta --extern ignore=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libignore-ca9c822506172d05.rmeta --extern itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-66e18d1a939ba1cd.rmeta --extern lazy_static=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/liblazy_static-014dca488f9f7695.rmeta --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/liblog-ffb0430a2e4c3fd2.rmeta --extern regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-5b62b80e23834394.rmeta --extern rustc_workspace_hack=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/librustc_workspace_hack-0ddfcc5fb3a9ae8c.rmeta --extern rustfmt_config_proc_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/librustfmt_config_proc_macro-ccc9a50f106ca471.so --extern serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-9b0a582ad2da8d4c.rmeta --extern serde_json=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde_json-575c758f47c1601b.rmeta --extern term=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libterm-d47930cf6ff2d399.rmeta --extern thiserror=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libthiserror-b2ff0d07581d967a.rmeta --extern toml=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtoml-ced06e0a80176a52.rmeta --extern unicode_segmentation=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libunicode_segmentation-753e77d677deb0b4.rmeta --extern unicode_width=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libunicode_width-fa39735c9cc400b9.rmeta --extern unicode_categories=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libunicode_categories-4f77959785227ff2.rmeta -Csymbol-mangling-version=v0 -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(parallel_compiler)' '--check-cfg=values(release)' '--check-cfg=values(no_btreemap_remove_entry)' '--check-cfg=values(crossbeam_loom)' '--check-cfg=values(span_locations)' '--check-cfg=values(rustix_use_libc)' -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.66.0-nightly
  (24100e3af
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] rustfmt_nightly test:false 7.903
Build completed unsuccessfully in 0:28:36
```

</details>

bors commented:
:hourglass: Testing commit 14de94aec59bcd535f690e255a950ca76fcf2d5a with merge 062de58b6ed12dd1f9bedbfed96ffa5a4e5db1ef...
<!-- homu: {"type":"BuildStarted","head_sha":"14de94aec59bcd535f690e255a950ca76fcf2d5a","merge_sha":"062de58b6ed12dd1f9bedbfed96ffa5a4e5db1ef"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3221464710/jobs/5269742284)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3221464710/jobs/5269742284","builder_name":"checks-actions"} -->

GuillaumeGomez committed:
Fix unclosed HTML tag in clippy doc

bors commented:
:pushpin: Commit d565200270d425ab2d72b06109314c786676f593 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle d565200270d425ab2d72b06109314c786676f593 -->
<!-- homu: {"type":"Approved","sha":"d565200270d425ab2d72b06109314c786676f593","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d565200270d425ab2d72b06109314c786676f593 with merge 36c8e291a617ae6bd4b8ff13c54c82862eaf0eec...
<!-- homu: {"type":"BuildStarted","head_sha":"d565200270d425ab2d72b06109314c786676f593","merge_sha":"36c8e291a617ae6bd4b8ff13c54c82862eaf0eec"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3222533601/jobs/5273198303)
Approved by: notriddle
Pushing 36c8e291a617ae6bd4b8ff13c54c82862eaf0eec to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"notriddle","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3222533601/jobs/5273198303"},"merge_sha":"36c8e291a617ae6bd4b8ff13c54c82862eaf0eec"} -->

rust-timer commented:
Finished benchmarking commit (36c8e291a617ae6bd4b8ff13c54c82862eaf0eec): [comparison URL](https://perf.rust-lang.org/compare.html?start=a6b7274a462829f8ef08a1ddcdcec7ac80dbf3e1&end=36c8e291a617ae6bd4b8ff13c54c82862eaf0eec&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a6b7274a462829f8ef08a1ddcdcec7ac80dbf3e1&end=36c8e291a617ae6bd4b8ff13c54c82862eaf0eec&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a6b7274a462829f8ef08a1ddcdcec7ac80dbf3e1&end=36c8e291a617ae6bd4b8ff13c54c82862eaf0eec&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 3.3%     | [2.7%, 4.0%] | 2         |
| Improvements ✅ <br /> (primary) | -3.2%    | [-4.6%, -2.1%] | 3         |
| Improvements ✅ <br /> (secondary) | -2.5%    | [-3.3%, -1.3%] | 26        |
| All ❌✅ (primary) | -3.2%    | [-4.6%, -2.1%] | 3         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a6b7274a462829f8ef08a1ddcdcec7ac80dbf3e1&end=36c8e291a617ae6bd4b8ff13c54c82862eaf0eec&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 3.1%     | [3.1%, 3.1%] | 1         |
| Regressions ❌ <br /> (secondary) | 3.2%     | [2.2%, 4.3%] | 2         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 3.1%     | [3.1%, 3.1%] | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*
